src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,802 @@
+
+	values[Anum_pg_dist_commit_transaction_transaction_id - 1] =
+		CStringGetTextDatum(transactionId);
+	values[Anum_pg_dist_commit_transaction_cluster_clock - 1] =
+		HeapTupleGetDatum(clockTuple);
+	values[Anum_pg_dist_commit_transaction_timestamp - 1] =
+		Int64GetDatum(transactionTimestamp);
+
+	/* open pg_dist_commit_transaction and insert new tuple */
+	Relation pgDistCommitTransaction =
+		table_open(DistCommitTransactionRelationId(), RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCommitTransaction);
+	HeapTuple heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	CatalogTupleInsert(pgDistCommitTransaction, heapTuple);
+
+	CommandCounterIncrement();
+
+	/* close relation and invalidate previous cache entry */
+	table_close(pgDistCommitTransaction, NoLock);
+}
+
+
+/*
+ * Initialize the shared memory clock value to the highest clock
+ * persisted. This will protect from any clock drifts.
+ */
+void
+InitClockAtBoot(void)
+{
+	if (creating_extension)
+	{
+		/* No catalog present yet */
+		return;
+	}
+
+	SpinLockAcquire(&logicalClockShmem->clockMutex);
+
+	/* Avoid repeated and parallel initialization */
+	if (logicalClockShmem->clockInitialized == CLOCKSTATE_INITIALIZED ||
+		logicalClockShmem->clockInitialized == CLOCKSTATE_INIT_INPROGRESS)
+	{
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+		return;
+	}
+
+	Assert(logicalClockShmem->clockInitialized == CLOCKSTATE_UNINITIALIZED);
+
+	/*
+	 * Set the flag before executing a distributed query
+	 * (else it might trigger this routine recursively.)
+	 */
+	logicalClockShmem->clockInitialized = CLOCKSTATE_INIT_INPROGRESS;
+
+	SpinLockRelease(&logicalClockShmem->clockMutex);
+
+	PG_TRY();
+	{
+		/* Start with the wall clock value */
+		uint64 epochValue = GetEpochTimeMs();
+
+		SpinLockAcquire(&logicalClockShmem->clockMutex);
+		SET_CLOCK(logicalClockShmem->clusterClockValue, epochValue, 0);
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+
+		/*
+		 * Version checking is a must for the below code as it accesses the
+		 * new catalog created in the new version.
+		 * Note: If we move this code to the top of the routine it will cause an
+		 * infinite loop with CitusHasBeenLoaded() calling this routine again.
+		 */
+		bool saveEnableVersionChecks = EnableVersionChecks;
+		EnableVersionChecks = true;
+
+		if (!CheckCitusVersion(NOTICE))
+		{
+			EnableVersionChecks = saveEnableVersionChecks;
+			return;
+		}
+
+		EnableVersionChecks = saveEnableVersionChecks;
+
+		/*
+		 * Select the highest clock value persisted in the catalog.
+		 *
+		 * SELECT i, MAX(j) FROM <tab>
+		 * WHERE i = (SELECT MAX(i) FROM <tab>)
+		 * GROUP BY i
+		 */
+		uint32 numCols = 2;
+		uint64 *results = ExecuteQueryAndReturnBigIntCols(
+			"SELECT (cluster_clock_value).logical, (cluster_clock_value).counter FROM "
+			"pg_catalog.pg_dist_commit_transaction ORDER BY 1 DESC, "
+			"2 DESC LIMIT 1;", numCols, SPI_OK_SELECT);
```

Comment:
We don't expect this table to be too big, we recommend the application to prune the table regularly, may be 1 hour worth of data is what we are expecting. Added a new index anyway and the plan shows

![image](https://user-images.githubusercontent.com/44680808/193732566-74f460f9-40d9-4dad-8308-9bdc4f77a850.png)

Reply:
>>If we do need an index, any reason to use SPI here instead of direcly accessing the catalog via heap scan's on the index, like we do with most -if not all- of our catalog tables

Not sure on how to get the effect of ORDER BY 1 DESC, 2 DESC LIMIT 1 from the direct index scan, any hints?

Reply:
> we recommend the application to prune the table regularly

I think there should be _one_ owner of the table. We cannot, as Citus, populate a table, and tell users to prune it every hour. This table grows with the rate of transactions, which is significant in certain environments There is something to say to not enable the clusterclock in those cases, but why do we put it in Citus core in that case?

Would there be any technical limitation for an application to maintain such a table them selfs if they need it? This is assuming we can safeguard to clock properties in Citus.

Reply:
Catalog is removed, so, this is not needed anymore